pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -147,91 +236,356 @@ func (p *Pod) PodSets() []kueue.PodSet {
+		}
+
+		podSets, err = p.constructGroupPodSets(p.list)
+		if err != nil {
+			if jobframework.IsUnretryableError(err) {
+				r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", err.Error())
+			}
+			return nil, err
+		}
+
+		if len(podSets) > 8 {
+			return nil, jobframework.UnretryableError(errMsgIncorrectGroupRoleCount)
+		}
+	}
+
+	wl := &kueue.Workload{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:  p.pod.GetNamespace(),
+			Labels:     map[string]string{},
+			Finalizers: []string{kueue.ResourceInUseFinalizerName},
+		},
+		Spec: kueue.WorkloadSpec{
+			PodSets:   podSets,
+			QueueName: jobframework.QueueName(p),
+		},
+	}
+
+	if p.groupName() == "" {
+		wl.Name = jobframework.GetWorkloadNameForOwnerWithGVK(p.pod.GetName(), p.GVK())
+		jobUid := string(p.Object().GetUID())
+		if errs := validation.IsValidLabelValue(jobUid); len(errs) == 0 {
+			wl.Labels[controllerconsts.JobUIDLabel] = jobUid
+		} else {
+			log.V(2).Info(
+				"Validation of the owner job UID label has failed. Creating workload without the label.",
+				"ValidationErrors", errs,
+				"LabelValue", jobUid,
+			)
+		}
+
+		// add the owner ref
+		if err := controllerutil.SetControllerReference(p.Object(), wl, c.Scheme()); err != nil {
+			return nil, err
+		}
+	} else {
+		wl.Name = p.groupName()
+		for _, pod := range p.list.Items {
+			if err := controllerutil.SetOwnerReference(&pod, wl, c.Scheme()); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return wl, nil
+}
+
+func (p *Pod) FindMatchingWorkloads(ctx context.Context, c client.Client) (*kueue.Workload, []*kueue.Workload, error) {
+	log := ctrl.LoggerFrom(ctx)
+	groupName := p.groupName()
+
+	if groupName == "" {
+		return jobframework.FindMatchingWorkloads(ctx, c, p)
+	}
+
+	// Find a matching workload first if there is one.
+	workload := &kueue.Workload{}
+	if err := c.Get(ctx, types.NamespacedName{Name: p.groupName(), Namespace: p.pod.GetNamespace()}, workload); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil, nil
+		}
+		log.Error(err, "Unable to get related workload")
+		return nil, nil, err
+	}
+
+	jobPodSets, err := p.constructGroupPodSets(p.list)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if p.equivalentToWorkload(workload, jobPodSets) {
+		return workload, []*kueue.Workload{}, nil
+	} else {
+		return nil, []*kueue.Workload{workload}, nil
+	}
+}
+
+func (p *Pod) equivalentToWorkload(wl *kueue.Workload, jobPodSets []kueue.PodSet) bool {
+	workloadFinished := apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadFinished)
+
+	if wl.GetName() != p.groupName() {
+		return false
+	}
+
+	if !workloadFinished && len(wl.Spec.PodSets) < len(jobPodSets) {
+		return false
+	}
+
+	for i := range wl.Spec.PodSets {
+		if i >= len(jobPodSets) {
+			return true
+		}
+		if !workloadFinished && wl.Spec.PodSets[i].Count < jobPodSets[i].Count {
+			return false
+		}
+		if wl.Spec.PodSets[i].Name != jobPodSets[i].Name {
+			return false
+		}
```

Comment:
Let's imagine there are 2 podsets, each with 1 pod.

If one of the pods is missing, the equivalency check should pass. I don't this code is letting that pass.

Reply:
If one pod is missing, this part will return true after the loop went over the first pod set:
https://github.com/kubernetes-sigs/kueue/blob/853a3d1235c8aaec9349629a2d8ac3b582211f7f/pkg/controller/jobs/pod/pod_controller.go#L573-L575

Reply:
if you have two podsets in the workload `{{Name: a, Count: 1}, {Name: b, Count: 1}}`, and the list of pods gives you `{{Name: b, Count: 1}}`, would this function return true? It doesn't look like it would.